### ISSUE TITLE
Implement torch.*_like tensor creation functions on sparse inputs

### ISSUE THREAD

**Issue by IvanYashchuk** [pytorch/pytorch#69822]
### 🚀 The feature, motivation and pitch

There should be a way in PyTorch to create a sparse tensor that is similar to one that is already created possibly with different values.
It would be great to get `torch.*_like` (`torch.empty_like`, `torch.rand_like`, `torch.randint_like`, etc.) working for sparse tensor inputs. All "like" tensor creation functions have the following behavior for the strided/dense tensors:
1. allocate the storage of the same size as `input.storage()` and copy shape, strides, device.
2. fill the allocated memory with values.

So for example `torch.randn_like` is a composition of `torch.empty_like` and elementwise inplace `normal_`.

For sparse inputs, the equivalent behavior would be:
1. allocate the values of the same size as `input.values()` and copy shape, indices, device.
2. fill the allocated memory (`.values()`) with values.

These tensor creation functions would preserve the sparsity pattern (copy the indices).

### Alternatives

There's no dense behavior equivalence if we allow sparse inputs for  `torch.*_like` functions. Meaning that 
`torch.randn_like(sparse_input).to_dense()` wouldn't be equal to `torch.randn_like(sparse_input.to_dense())`.
Because of this @cpuhrsch proposed to put new functionality under `torch.sparse` namespace. However, it would unnecessarily complicate existing code even in the PyTorch codebase moving the dense/sparse dispatching from the PyTorch Dispatcher into the outside world. All occurrences of "like" functions would need to be replaced with a conditional on the `.is_sparse()` of the input.

**Comment by vadimkantorov** [comment]
One other avenue:

NumPy now accepts `like=` argument to their factory methods: https://numpy.org/doc/stable/reference/generated/numpy.full.html

I had brought this up a few years back, but it didn't gather support I think

**Comment by pearu** [comment]
Nit: To be exact, `torch.randn_like(dense_input)` would not be equal to `torch.randn_like(dense_input)` because of random number generation. So "Meaning that torch.randn_like(sparse_input).to_dense() wouldn't be equal to torch.randn_like(sparse_input.to_dense())." does not make sense as it is.

A more sensible question is, should `torch.ones_like(sparse_input).to_dense()` be equal to `torch.ones_like(sparse_input.to_dense())`, or not?

According to pytorch docs, we have this definition for like-methods:
> torch.foo_like(input) is equivalent to torch.foo(input.size(), dtype=input.dtype, layout=input.layout, device=input.device)

that cannot be applied to sparse layouts (as the r.h.s. does not support specifying indices) without generalizations.
(Btw, storage consumption of `torch.foo_like(input)` may be much create than of `torch.foo(input.size(), dtype=input.dtype, layout=input.layout, device=input.device)` when `input` is a view of some larger tensor, but that storage-inefficency is another issue..)

For strided tensors, the like-methods have the following (undocumented?) property:
> torch.foo_like(input).stride() == input.stride()

which would translate for sparse tensor cases as follows:
> torch.foo_like(coo_input).indices() == coo_input.indices()
> torch.foo_like(csr_input).crow_indices() == csr_input.crow_indices()
> torch.foo_like(csr_input).col_indices() == csr_input.col_indices()

because both strides and sparse tensor indices describe how the values in the tensor storage map to a mathematical tensor representation, that is, a mapping of tensor element indices to the corresponding values. This view of the like-methods corresponds to the proposal here.

From the storage-efficiency perspective, like-methods ought to preserve input indices. Curiously, in the case of strided tensors, preserving input indices data (as specified by shape and strides), that leads to storage inefficiency.

**Comment by IvanYashchuk** [comment]
> Nit: To be exact, torch.randn_like(dense_input) would not be equal to torch.randn_like(dense_input) because of random number generation. So "Meaning that torch.randn_like(sparse_input).to_dense() wouldn't be equal to torch.randn_like(sparse_input.to_dense())." does not make sense as it is.

For random number generation, of course, the state needs to be reset for comparison:
```python
import torch

dense = torch.ones(3, 3)

torch.manual_seed(1)
sparse1 = torch.randn_like(dense).to_sparse()

torch.manual_seed(1)
sparse2 = torch.sparse_coo_tensor(sparse1.indices(), torch.empty_like(sparse1.values()), sparse1.size())
sparse2._values().normal_()

print(torch.allclose(sparse1.to_dense(), sparse2.to_dense()))
# True
```

**Comment by cpuhrsch** [comment]
@IvanYashchuk - the main motivation here is, as far as I understand, that our test code will need to be changed if we don't support them. These changes are seen as more costly than introducing these new functions. Could you post a few pointers to these call sites?

**Comment by pearu** [comment]
Here's an idea of applying torch operations, including factory methods, to sparse tensors with preserved sparsity.

Define
```python
def vapply(input, op, *args, **kwargs):
    if input.layout == torch.strided:
        return op(input, *args, **kwargs)
    elif input.is_sparse():
        return torch.sparse_coo_tensor(input._indices(), vapply(input._values(), op, *args, **kwargs), input.shape,
                  // dtype, device are defined by vapply result,
                  requires_grad=input.requires_grad)
    elif input.is_sparse_csr():
        return torch.sparse_csr_tensor(input.crow_indices(), input.col_indices(),
                   vapply(input.values(), op, *args, **kwargs), input.shape,
                   // dtype, device are defined by vapply result,
                   requires_grad=input.requires_grad)
    else:
        raise NotImplementedError(f'vapply for {input.layout=}')
```
so that we can apply any torch operations to tensors values
```python
vapply(coo_input, torch.randn_like)   # randn_like for sparse input
vapply(csr_input, torch.cos)          # cos applied to csr tensor values
```
I suggest that the function `vapply` would live in `torch` namespace as it can be applied to strided inputs as well. Also, it would be a Tensor method so that one can use:
```python
coo_input.vapply(torch.randn_like)
csr_input.vapply(torch.cos)
```
What do you think?

**Comment by cpuhrsch** [comment]
Hm, I think this could be something in torch.sparse. I also think this might be difficult to remove at some point, if we assume that MaskedTensor is the desired replacement for masked semantics. How do we best deal with the "unspecified value" problem in this case?

**Comment by pearu** [comment]
Re "unspecified value" problem: it is not a problem for `vapply` because it acts on specified values only.

I consider `vapply` as a utility function that hides users/developers dealing with indices as these are just passed on to the tensor constructor. `vapply` is not meant to resolve the problems that MaskedTensor tackles.

Re native functions: yes, implementing `vapply` in C++ may be difficult but possible as functions can be passed to `vapply` as template parameters. A quick workaround would be to define `vapply` as a CPP-macro..

Re dense layout support: notice that `vapply` as given above already supports strided tensors.

Yes, using `vapply` approach in testing sites to enable layout transparency for certain operators (like-facories, unary functions) makes sense. In fact, this would tackle the underlying problem of this issue: how to unify strided and sparse tensors so that existing tests could be applied to sparse tensors without changing the tests too much.

**Comment by IvanYashchuk** [comment]
> 1. Introduce new functions within torch.sparse

If we begin to choose this option here and for other functions as well, there is a risk of ending up with a copy of `torch.` namespace in the `sparse` module. It feels wrong to move the dispatching task from the PyTorch Dispatcher to the user, especially when the function signature doesn't contain any sparse related details and would be an exact copy of the existing function.

> 3. Defer users to a manual workaround using indices and values.

Choosing this option leads to a question why do we even have a container Tensor that stores indices and values. We could have a sparse module working directly with indices and values Tensors without putting them together in a struct, similarly as done in the [`pytorch_sparse`](https://github.com/rusty1s/pytorch_sparse) project. If this option becomes recommended then the support for elementwise functions and some other functions in the `torch` namespace should be reverted. If someone wants to get the absolute values then use
```python
t = sparse_coo_tensor(<...>)
o = sparse_coo_tensor(t.indices(), torch.abs(t.values()), t.size())
```
and repeat the same recommendation for other functions making COO Tensor useless.

> There are a couple ways in which we can view these *_like factory functions that might get us out of this situation without violating our to_dense constraint.
a. Use MaskedTensor.

Why is MaskedTensor allowed to be passed to `torch.ones_like` and sparse tensor is not? Why does the MaskedTensor doesn't have a "to_dense constraint"? Is it because the MaskedTensor prototype is implemented using a new Tensor subclassing mechanism (`__torch_dispatch__`) and sparse subclassing is marked with a "layout" attribute? The layout is just a marker that doesn't have a lot of utility, it's not even is used in `.is_sparse()`, `is_sparse_csr()` methods, or in `.to()`, it could be removed easily (it's in "beta") and little code would break.

**Comment by pearu** [comment]
I agree in that having two namespaces torch and torch.sparse is confusing when torch namespace contains operations that support sparse inputs, so we cannot say that torch namespace is for strided-only operations.
However, we should make clear that the purpose of `torch.sparse` namespace is to collect those operations that treat sparse tensors as masked tensors. Operations in the torch namespace support sparse inputs with the assumptions that these operations assume that unspecified elements are treated as zeros.

> Why is MaskedTensor allowed to be passed to torch.ones_like and sparse tensor is not?

This is because the semantics of ones_like is clear for MaskedTensor (return a tensor with all elements being 1) while it is ambiguous for sparse tensors that treat unspecified values as zeros: according to the suggestion of this issue, the result of `ones_like` on sparse tensor would not be a "tensor with all elements being 1".

> Why does the MaskedTensor doesn't have a "to_dense constraint"? Is it because the MaskedTensor prototype is implemented using a new Tensor subclassing mechanism (__torch_dispatch__) and sparse subclassing is marked with a "layout" attribute? 

These are implementation details that I think are irrelevant when discussing semantics.

> The layout is just a marker that doesn't have a lot of utility, it's not even is used in .is_sparse(), is_sparse_csr() methods

This is not completely true, see is_sparse_csr in torch/include/c10/core/TensorOptions.h

**Comment by cpuhrsch** [comment]
> 



> > 1. Introduce new functions within torch.sparse
> 
> If we begin to choose this option here and for other functions as well, there is a risk of ending up with a copy of `torch.` namespace in the `sparse` module. It feels wrong to move the dispatching task from the PyTorch Dispatcher to the user, especially when the function signature doesn't contain any sparse related details and would be an exact copy of the existing function.
> 
> > 3. Defer users to a manual workaround using indices and values.
> 
> Choosing this option leads to a question why do we even have a container Tensor that stores indices and values. We could have a sparse module working directly with indices and values Tensors without putting them together in a struct, similarly as done in the [`pytorch_sparse`](https://github.com/rusty1s/pytorch_sparse) project. If this option becomes recommended then the support for elementwise functions and some other functions in the `torch` namespace should be reverted. If someone wants to get the absolute values then use

We don't have to do this with all functions that are contentious from here on out. This doesn't have to be an "all or nothing" decision for *all* future discussion, but we should certainly step through the pros and cons, which are different for other categories of functions. Ideally we'd be able to hide the details of our storage layouts from users, unless they explicitly ask for them. Here we are however in a situation where the storage details matter, since we want to avoid densification and apply them to the materialized values directly.

As a technical nit, pytorch_sparse still has a [SparseTensor](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/tensor.py#L12) with methods and functions. I know it says it on their README that it's decomposed, but there's a lot of operations that are direct methods. For example see [the reductions](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/reduce.py#L86-L89). It's even JIT-able, supports dense dimensions, COO, CSR and CSC.

> 
> ```python
> t = sparse_coo_tensor(<...>)
> o = sparse_coo_tensor(t.indices(), torch.abs(t.values()), t.size())
> ```
> 
> and repeat the same recommendation for other functions making COO Tensor useless.

Indeed in cases where we can't maintain the to_dense constraint, this is a valid workaround. Another workaround would be to use a MaskedTensor, which can explicitly restrict the computation to materialized values only.

> 
> > There are a couple ways in which we can view these *_like factory functions that might get us out of this situation without violating our to_dense constraint.
> > a. Use MaskedTensor.
> 
> Why is MaskedTensor allowed to be passed to `torch.ones_like` and sparse tensor is not? Why does the MaskedTensor doesn't have a "to_dense constraint"? Is it because the MaskedTensor prototype is implemented using a new Tensor subclassing mechanism (`__torch_dispatch__`) and sparse subclassing is marked with a "layout" attribute? The layout is just a marker that doesn't have a lot of utility, it's not even is used in `.is_sparse()`, `is_sparse_csr()` methods, or in `.to()`, it could be removed easily (it's in "beta") and little code would break.

torch.ones_like will produce a MaskedTensor when given a MaskedTensor with the exact same storage properties. If the MaskedTensor is backed by a sparse data and sparse mask then the result of ones_like will also be a MaskedTensor backed by sparse data and sparse mask. For SparseTensor we can't maintain this constraint without violating the to_dense contract.

**Comment by cpuhrsch** [comment]
@IvanYashchuk @pearu 

What do you guys think about treating these ```*_like``` factory functions as memory allocators and calling it a day? I'd not want to apply this reasoning to other functions, but for factory functions it seems fine, because they're so specific to the storage details. For torch.Tensor they preserve the strides even if it means allocating a much larger Tensor first. Their real purpose is to produce a Tensor that's exactly like the one given, except using a different allocator for the values. This aligns with your proposal Ivan.

**Comment by pearu** [comment]
> What do you guys think about treating these `*_like` factory functions as memory allocators and calling it a day?

Yes, treating `*_like` factory functions as memory allocators makes sense. To be specific, we would have the following invariant (which can be used as a definition for the implementation):
```python
torch.*_like(coo, ...) == torch.sparse_coo_tensor(coo._indices(), torch.*_like(coo._values(), ...), coo.shape)
``
(notice that `coo` may be coalesced or not). A similar invariant exists for CSR tensors.
